cordapp-cpk/src/test/resources/external-cordapp/corda-platform-cordapp/build.gradle

```diff
@@ -0,0 +1,26 @@
+plugins {
+    id 'net.corda.plugins.cordapp-cpk'
+}
+
+cordapp {
+
+    targetPlatformVersion = platform_version.toInteger()
+    minimumPlatformVersion = platform_version.toInteger()
+
+    contract {
+        name = 'Corda platform CorDapp'
+        versionId = cordapp_contract_version.toInteger()
+        licence = 'Test-Licence'
+        vendor = 'R3'
+    }
+}
+
+dependencies {
+//    cordaProvided project(":corda-api")
```

Comment:
for some reason, if I uncomment this the `:external-cordapp-transitive-dependnecy` cpk is not included in the assembled CPB

Reply:
Perhaps because you haven't published `corda-api` to your "Test" repository?

Reply:
I haven't but I don't understand why that would be an issue, that is a project dependency, not an external dependency

Reply:
> I haven't but I don't understand why that would be an issue, that is a project dependency, not an external dependency

Because once published into a Maven repository, the transitive dependency on `corda-api` becomes:
```xml
<dependency>
    <groupId>net.corda</groupId>
    <artifactId>corda-api</artifactId>
    <version>5.0.0</version>
</dependency>
```
which Gradle no longer recognises as `project(':corda-api')`.

Reply:
shouldn't it then throw an exception saying that such a dependency cannot be found in any Maven repo? 

it silently "works" instead and it excludes the CPK of `:external-cordapp-transitive-dependency` which I don't understand how is related to `project(':corda-api')`

Reply:
Have you checked the output of the test Gradle projects for any warnings? I believe the "transitive CPK" mechanism _does_ warn if it cannot find a CPK's companion POM, which also contains the CPKs `cordaProvided` dependencies. However, Gradle cannot _fail_ if there's a problem with the companion POM because it might belong to a third-party CPK and so not be fixable by the current CorDapp developer.

Reply:
as far as I know, if a transitive dependency cannot be found anywhere, the dependency resolution fails even if you, as a developer, cannot do anything about it (because it's not a direct dependency). I don't think we should have a different behavior...

For the warning.. I don't see any warning apart from a few 

```
[ant:signjar] 
[ant:signjar] Warning: 
[ant:signjar] The signer's certificate is self-signed.
```

which I assume to be related to the fact that I'm using the dev cert for signing

Reply:
I mean, if I publish a library that depends on `corda-base` and I consume it from a project that is missing the R3 artifactory repo from the `repositories` list, Gradle fails complaining that `corda-base` cannot be found anywhere, I don't think there's anything wrong with it

Reply:
I found the warning though
```
Resource missing. [HTTP GET: https://repo.maven.apache.org/maven2/net/corda/corda-api/5.0.0/corda-api-5.0.0.pom]
CorDapp has unresolved dependencies:
- net.corda:corda-api:5.0.0
Cannot resolve CPK companion artifact 'com.example.external-cordapp:external-cordapp.corda.cpk:1.2.1' - SKIPPED
```

Reply:
I do not understand the last line though.. I'm investigating it

Reply:
I'm fairly sure there are warnings like these somewhere:
```
Configuring cordapp-cpb as a cordapp
Resource missing. [HTTP GET: https://repo.maven.apache.org/maven2/net/corda/corda-api/5.0.0/corda-api-5.0.0.pom]
CorDapp has unresolved dependencies:
- net.corda:corda-api:5.0.0
Cannot resolve CPK companion artifact 'com.example.external-cordapp:external-cordapp.corda.cpk:1.2.1' - SKIPPED
```

Reply:
The point is that the `cordapp-cpk` plugin found the companion POM for
```
com.example.external-cordapp:external-cordapp.corda.cpk:1.2.1
```
but couldn't resolve it because
```
net.corda:corda-api:5.0.0
```
had not been published to your "Test" repository.

Reply:
ok, but why is it excluding the dependencies that could be resolved though? `:external-cordapp-transitive-dependency` has been published in the test repo, even if it's `cordaProvided` dependency, `corda-api` was not... I do not think it is a good idea to skip `corda-api` because if it cannot not be found but I especially do not understand why `external-cordapp-transitive-dependency` is also skipped

Reply:
> as far as I know, if a transitive dependency cannot be found anywhere, the dependency resolution fails even if you, as a developer, cannot do anything about it (because it's not a direct dependency).

This transitive resolution is being performed by the `CordappDependencyCollector` as it tries to resolve the "companion" POMs. The "companion" POM is an extra "helper" artfifact that I make a _best effort_ to publish using _Dark Gradle Magic_, and so I cannot _demand_ that every published CPK will have one.

Reply:
imho you should require that if it is required for transitive dependencies resolution, otherwise you would end up assebling CPB or installing CPKs that cannot be started